### ISSUE TITLE
Diamond task fusion

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1964]
We sometimes fuse linear graphs.  This reduces scheduler overhead and minimizes the chance of errant data transfer.

Sometimes we encounter parts of dependency graphs that look like the following:

```
  o
 / \
o   o
|   |
o   |
 \ /
  o
```

Normally we don't fuse these graphs because they have potential for parallelism.  However, if we find ourselves in a situation where scheduler overhead causes performance issues, we may desire to fuse more aggressively.  Motifs like the diamond pattern above occur frequently and, if we are comfortable fusing them, can make a significant impact in some workloads.  

Notably, this came up in dask-glm (where it is a bottleneck) and other dask.array workloads that use the same array in mathematical expressions multiple times, such as in the following example:

    y = x ** 2 + 2 * x

So it would be nice to be able to optionally fuse these.

```python
def test_diamond_fusion():
    dsk = {'a': 1, 'b': (inc, 'a'), 'c': (dec, 'a'), 'd': (add, 'b', 'c')}
    expected = {'d-b-c': (add, (inc, 'a'), (dec, 'a')), 'a': 1}  # or maybe we also fuse 'a'?
    result = diamond_fuse(dsk, [])
    assert expected == result
```

@sklam, @eriknw and I chatted about this a bit after discussing dask-glm algorithms.  @sklam brought up the following links:

- https://github.com/numba/numba/blob/master/numba/controlflow.py#L338
- https://pdfs.semanticscholar.org/2e3b/19aa1df44df18435900d3deeb11a547d193e.pdf
- https://github.com/sklam/codetrace/blob/master/codetrace/controlflow.py#L500

Some things to note:

-  We need this to be in linear time
-  We already have a nice fuse operation

**Comment by pitrou** [comment]
What strikes me about this particular example is that expressions such as `x ** 2 + 2 * x` would really benefit from optimizing using Numba or numexpr:
```
>>> %timeit x ** 2 + 2 * x
100 loops, best of 3: 5.9 ms per loop
>>> %timeit numexpr.evaluate("x ** 2 + 2 * x")
1000 loops, best of 3: 1.46 ms per loop
>>> %timeit f(x)
1000 loops, best of 3: 1.39 ms per loop
```

**Comment by mrocklin** [comment]
Yes, I think that after we start fusing more then another optimization to replace tasks like `(add, (inc, 'a'), (dec, 'a'))` with `(numba.jit(lambda a: x + 1 + x - 1), 'a')` would be nice.  At the moment though we're far more concerned with the ~200us cost on the scheduler than the 5ms cost on the workers.

**Comment by mrocklin** [comment]
Looking at this a bit more deeply I think that the situation can be simplified a bit, particularly if we perform a linear `fuse` ahead of time.  There are only so many possible output configurations with small out-degree (anything larger can be immediately dismissed).

**Comment by eriknw** [comment]
I've been thinking (and doodling) on this as well, and it is indeed convenient to assume `fuse` has already been performed.  I also think "single input, single output" is too restrictive--we only need "single output".  Essentially, the desired optimization is to combine reductions that are less than a specified width (i.e., "width = num_nodes / height" is a measure of parallelism of the merged tasks).

I'll probably start coding something up in a few days.  @mrocklin, I'm happy to discuss this in person, or here as time permits.  I also think this problem is nuanced enough that there could be value in having multiple attempts or approaches.

**Comment by mrocklin** [comment]
Thinking about "single output only" is an interesting idea.  I hadn't thought of thta.

I should be significantly more free and still in Austin on Friday if you're around for whiteboarding.

**Comment by jcrist** [comment]
Fixed by #1964.

**Comment by jakirkham** [comment]
Looks to be solved in PR ( https://github.com/dask/dask/pull/1979 ).

cc @quasiben (for awareness)
